pkg/agent/controller/proxy/controller.go

```diff
@@ -0,0 +1,121 @@
+package proxy
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Reconcile struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	baseSvcCache *BaseSvcCache
+}
+
+func (r *Reconcile) ReconcileService(req ctrl.Request) (ctrl.Result, error) {
+	var svc corev1.Service
+	ctx := context.Background()
+	err := r.Get(ctx, req.NamespacedName, &svc)
+	if client.IgnoreNotFound(err) != nil {
+		klog.Errorf("Failed to get service: %v, err: %s", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	baseSvcID := genSvcID(req.Namespace, req.Name)
+	old := r.baseSvcCache.Get(baseSvcID)
+
+	if apierrors.IsNotFound(err) {
+		if err := r.deleteService(ctx, old); err != nil {
+			klog.Errorf("Failed to delete service: %v, err: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		klog.Infof("Success delete service: %v", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	if old != nil {
+		if err := r.updateService(ctx, &svc, old); err != nil {
+			klog.Infof("Failed to update service: %v, err: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		klog.Infof("Success update service: %v", req.NamespacedName)
+	} else {
+		if err := r.createService(ctx, &svc); err != nil {
+			klog.Infof("Failed to create service: %v, err: %s", req.NamespacedName, err)
+		}
+		klog.Infof("Success create service: %v", req.NamespacedName)
```

Comment:
create failed and will also print success message?
when failed need to return err